File: 19. Rotate Matrix k Times.py
"""Topic: Rotate matrix elements k times (flattened)"""
from collections import deque
def rotate_k(A,k):
    flat=[v for r in A for v in r]
    k%=len(flat)
    dq=deque(flat)
    dq.rotate(k)
    it=iter(dq)
    n=len(A); m=len(A[0])
    B=[ [next(it) for _ in range(m)] for _ in range(n) ]
    return B

if __name__=='__main__':
    A=[[1,2,3],[4,5,6],[7,8,9]]
    for r in rotate_k(A,2): print(' '.join(map(str,r)))

# TC: O(n*m)
# SC: O(n*m)

File: 19. Rotate Matrix k Times.cpp
// Topic: Rotate matrix elements k times in circular fashion (on flattened array)
#include <bits/stdc++.h>
using namespace std;
int main()
{
    vector<vector<int>> A = {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}};
    int n = A.size(), m = A[0].size();
    int k = 2;
    vector<int> flat;
    for (auto& r : A)
        for (auto v : r) flat.push_back(v);
    int len = flat.size();
    k %= len;
    if (k < 0)
        k += len;
    rotate(flat.begin(), flat.end() - k, flat.end());
    vector<vector<int>> B(n, vector<int>(m));
    for (int i = 0, idx = 0; i < n; ++i)
        for (int j = 0; j < m; ++j) B[i][j] = flat[idx++];
    for (auto& r : B)
    {
        for (auto v : r) cout << v << " ";
        cout << "\n";
    }
}

// TC: O(n*m)
// SC: O(n*m)
